[lldb] Introduce internal stop hooks

yln opened the pull request:
Introduce the concept of internal stop hooks.
These are similar to LLDB's internal breakpoints:
LLDB itself will add them and users of LLDB will
not be able to add or remove them.

This change adds the following 3
independently-useful concepts:
* Maintain a list of internal stop hooks that will be populated by LLDB and cannot be added to or removed from by users.  They are managed in a separate list in `Target::m_internal_stop_hooks`.
* `StopHookKind:CodeBased` and `StopHookCoded` represent a stop hook defined by a C++ code callback (instead of command line expressions or a Python class).
* Stop hooks that do not print any output can now also suppress the printing of their header and description when they are hit via `StopHook::GetSuppressOutput`.

Combining these 3 concepts we can model "internal
stop hooks" which serve the same function as
LLDB's internal breakpoints: executing built-in,
LLDB-defined behavior, leveraging the existing
mechanism of stop hooks.

This change also simplifies
`Target::RunStopHooks`.  We already have to
materialize a new list for combining internal and
user stop hooks.  Filter and only add active hooks to this list to avoid the need for "isActive?"
checks later on.

yln committed:
[lldb] Introduce internal stop hooks

Introduce the concept of internal stop hooks.
These are similar to LLDB's internal breakpoints:
LLDB itself will add them and users of LLDB will
not be able to add or remove them.

This change adds the following 3
independently-useful concepts:
* Maintain a list of internal stop hooks that will
  be populated by LLDB and cannot be added to or
  removed from by users.  They are managed in a
  separate list in
  `Target::m_internal_stop_hooks`.
* `StopHookKind:CodeBased` and `StopHookCoded`
  represent a stop hook defined by a C++ code
  callback (instead of command line expressions or
  a Python class).
* Stop hooks that do not print any output can now
  also suppress the printing of their header and
  description when they are hit via
  `StopHook::GetSuppressOutput`.

Combining these 3 concepts we can model "internal
stop hooks" which serve the same function as
LLDB's internal breakpoints: executing built-in,
LLDB-defined behavior, leveraging the existing
mechanism of stop hooks.

This change also simplifies
`Target::RunStopHooks`.  We already have to
materialize a new list for combining internal and
user stop hooks.  Filter and only add active hooks
to this list to avoid the need for "isActive?"
checks later on.

yln committed:
[NFC][lldb] Fix outdated comment

DavidSpickett commented:
I have used stop hooks as a user but never had to deal with their implementation. So my first question here is "but why?".

The strongest point here seems to be:
> Combining these 3 concepts we can model "internal
> stop hooks" which serve the same function as
> LLDB's internal breakpoints: executing built-in,
> LLDB-defined behavior, leveraging the existing
> mechanism of stop hooks.

And maybe those who have worked on them will think that is enough to make this change.

Is there a bigger motivation for this?

jimingham commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -1433,7 +1444,7 @@ class Target : public std::enable_shared_from_this<Target>,
   private:
     StringList m_commands;
-    // Use CreateStopHook to make a new empty stop hook. The GetCommandPointer
+    // Use CreateStopHook to make a new empty stop hook. The SetActionFromString
```
This wasn't grammatical originally, but...
Maybe:

Use SetActionFromString to fill it with commands

yln committed:
[lldb] Add option to show internal stop hooks

Add option to the existing "print list of stop
hooks" command to show the internal stop hooks
instead:
```
target stop-hook list --internal
```

yln committed:
[lldb] Unify printing of internal and user stop hooks

Unify the printing of internal and user
(non-internal) stop hooks to make it more similar
to how we treat it for breakpoints.

This means replacing the `GetNumStopHooks()` and
`GetStopHookAtIndex()` with a copy of the stop
hooks (in C++20, we can change this to a constant
view). These functions were ever only used to
iterate the entire collection of stop hooks and
the implementation of `GetStopHookAtIndex()` turns
this into a `N^2` operation.

This change also allows us to cleanup a bit of
code at the call site.

yln commented:
> I also think for debugging purposes, we'll want to have `target stop-hook list -I`.

I've added `target stop-hook list --internal`:

<img width="382" height="166" alt="Screenshot 2025-10-23 at 4 33 35 PM" src="https://github.com/user-attachments/assets/b2c0ba96-91e5-40b8-9c79-b6028b1e9578" />

yln commented on lldb/source/Commands/CommandObjectBreakpoint.cpp:
```diff
@@ -1114,9 +1114,7 @@ class CommandObjectBreakpointList : public CommandObjectParsed {
-            "List some or all breakpoints at configurable levels of detail.",
-            nullptr) {
-    CommandArgumentData bp_id_arg;
+            "List some or all breakpoints at configurable levels of detail.") {
```
Yes, this was just a cleanup.

jimingham commented on lldb/source/Target/Target.cpp:
```diff
@@ -3098,6 +3107,23 @@ void Target::SetAllStopHooksActiveState(bool active_state) {
   }
 }
 
+// FIXME:  Ideally we would like to return a `const &` (const reference) instead
```
Are you sure that using the iterator of the map will preserve the order of the stop-hooks in this listing?  I'm not concerned about internal stop-hooks, but it wouldn't be good if you added a bunch of stop hooks, then deleted some, and then the order of the stop-hooks in the listing got scrambled.
Might be good to add a test to ensure that this ordering is maintained to ensure that.

yln replied:
Added new test to pin down the behavior: 27b1441d3e27a12a4
Made sure it goes red when order is messed up, for example:
```
for (auto &[_, hook] : llvm::reverse(m_stop_hooks))
    stop_hooks.push_back(hook);
```

yln committed:
[lldb] Remove unimplemented function

Remove unused and unimplemented function
`GetStopHookSize()` declared in `Target.h`.

yln committed:
[lldb] Add test for stop hook ordering

Add test for stop hook user ID assignment,
ordering, and printing.

jimingham reviewed (approved):
LGTM
